File: core/analytics/src/androidMain/kotlin/com/gasguru/core/analytics/MixpanelAnalyticsHelper.kt
package com.gasguru.core.analytics

import com.mixpanel.android.mpmetrics.MixpanelAPI
import org.json.JSONObject

class MixpanelAnalyticsHelper(
    private val mixpanel: MixpanelAPI,
) : AnalyticsHelper {

    override fun logEvent(event: AnalyticsEvent) {
        val properties = JSONObject()
        properties.put(AnalyticsEvent.ParamKeys.CATEGORY, event.category)
        event.extras.forEach { param ->
            properties.put(param.key, param.value)
        }
        mixpanel.track(event.type, properties)
    }

    override fun updateSuperProperties(properties: Map<String, Any>) {
        val jsonProperties = JSONObject()
        properties.forEach { (key, value) -> jsonProperties.put(key, value) }
        mixpanel.registerSuperProperties(jsonProperties)
    }
}

File: iosApp/iosApp/MixpanelAnalyticsHelper.swift
import ComposeApp
import Mixpanel

final class MixpanelAnalyticsHelperIos: NSObject, AnalyticsHelper {

    override init() {
        let token = AnalyticsModuleIosKt.getMixpanelToken()
        Mixpanel.initialize(options: MixpanelOptions(token: token, trackAutomaticEvents: false))
    }

    func logEvent(event: AnalyticsEvent) {
        var properties: [String: MixpanelType] = ["category": event.category]
        for param in event.extras {
            properties[param.key] = param.value
        }
        Mixpanel.mainInstance().track(event: event.type, properties: properties)
    }

    func updateSuperProperties(properties: [String: Any]) {
        let mixpanelProps = properties.reduce(into: [String: MixpanelType]()) { result, pair in
            if let value = pair.value as? MixpanelType {
                result[pair.key] = value
            }
        }
        Mixpanel.mainInstance().registerSuperProperties(mixpanelProps)
    }
}
